manager/dispatcher/dispatcher.go

```diff
@@ -237,27 +234,19 @@ func (d *Dispatcher) Stop() error {
 	d.cancel()
 	d.mu.Unlock()
 	d.nodes.Clean()
-	// wait for all handlers to finish their raft deals, because manager will
-	// set raftNode to nil
-	d.wg.Wait()
 	return nil
 }
 
-func (d *Dispatcher) addTask() error {
+func (d *Dispatcher) canAddTask() error {
```

Comment:
This function is very racy in nature. By the time you check `canAddTask`, the condition may no longer be true. Some sort of lock needs to be held over the course of the transaction.